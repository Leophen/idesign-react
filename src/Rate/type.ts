export interface RateProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 固定分值（受控）
   */
  value?: number;
  /**
   * 默认分值（非受控）
   * @default 0
   */
  defaultValue?: number;
  /**
   * 是否只读
   * @default false
   */
  readonly?: boolean;
  /**
   * 是否可清除
   * @default false
   */
  allowClear?: boolean;
  /**
   * 是否支持半星
   * @default false
   */
  allowHalf?: boolean;
  /**
   * 最大分值
   * @default 5
   */
  count?: number;
  /**
   * 选中颜色
   * @default #f5db4d
   */
  activeColor?: string;
  /**
   * 未选中颜色
   * @default #e5e6ea | #333336
   */
  voidColor?: string;
  /**
   * 选中图标
   * @default StarFill
   */
  activeIcon?: string;
  /**
   * 未选中图标
   * @default StarFill
   */
  voidIcon?: string;
  /**
   * 选择评分时触发
   */
  onChange?: (val: number) => void;
}
