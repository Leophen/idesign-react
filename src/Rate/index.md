---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 17
---

# Rate 评分

用于评分打星的组件。

## 基本用法

```tsx
import React, { useState } from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(3)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Rate />
      <h4>有默认值</h4>
      <Rate defaultValue={value} />
      <h4>有受控值</h4>
      <Rate value={value} />
      <h4>一般用法</h4>
      <Rate
        value={value}
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 只读状态

可通过 `readonly` 属性控制为只读：

```tsx
import React from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  return (
    <Rate value={3} readonly />
  )
}

export default App
```

## 可清除分值

可通过 `allowClear` 属性控制为再次点击分值可清除的模式：

```tsx
import React from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  return (
    <Rate allowClear />
  )
}

export default App
```

## 半星状态

可通过 `allowHalf` 属性控制为半星模式：

```tsx
import React, { useState } from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(2.5)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <Rate allowHalf />
      <br />
      <Rate
        value={value}
        allowHalf
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 最大分值

可通过 `count` 属性控制最大分值：

```tsx
import React from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  return (
    <Rate count={10} />
  )
}

export default App
```

## 自定义颜色

可通过 `activeColor` 和 `voidColor` 属性控制选中和未选中颜色：

```tsx
import React from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  return (
    <Rate activeColor="#427FEA" voidColor="#E06B67" />
  )
}

export default App
```

## 自定义图标

可通过 `activeIcon` 和 `voidIcon` 属性控制选中和未选中颜色：

```tsx
import React from 'react'
import { Rate } from 'idesign-react'

const App = () => {
  return (
    <Rate activeIcon="BellFill" voidIcon="Bell" />
  )
}

export default App
```

<API />
