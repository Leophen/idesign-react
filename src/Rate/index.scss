.i-rate {
  display: flex;
  align-items: center;
  &.i-rate__readonly {
    .i-rate-star {
      cursor: default;
      &:active {
        transform: unset;
      }
    }
  }
}

.i-rate-star {
  position: relative;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    color: var(--i-bg-back);
  }
  &.i-rate-star__active {
    .i-icon {
      color: #f2cd5e;
    }
  }
  &:active {
    transform: scale(1.2);
  }
  &:not(:last-of-type) {
    padding-right: 8px;
  }
}

.i-rate-star__first {
  position: absolute;
  top: 0;
  left: 0;
  width: 10px;
  overflow: hidden;
}
