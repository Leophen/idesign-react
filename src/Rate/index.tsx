import React, { useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import useDefault from '../hooks/useDefault';
import tinycolor from 'tinycolor2';
import { RateProps } from './type';

const Rate: React.FC<RateProps> = (props) => {
  const {
    className,
    style,
    value,
    defaultValue = 0,
    readonly = false,
    allowClear = false,
    allowHalf = false,
    count = 5,
    activeColor,
    voidColor,
    activeIcon = 'StarFill',
    voidIcon = 'StarFill',
    onChange = () => { },
    ...restProps
  } = props;

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);
  const [hoverValue, setHoverValue] = useState(value || defaultValue)
  const [ifHover, setIfHover] = useState(false)

  const handleEnterRate = () => {
    !readonly && setIfHover(true)
  }

  const handleLeaveRate = () => {
    !readonly && setIfHover(false)
  }

  const handleEnterRateItem = (index: number) => {
    const step = !allowHalf ? 1 : 0.5
    !readonly && setHoverValue(index + step)
  }

  const handleDownRateItem = (index: number) => {
    const step = !allowHalf ? 1 : 0.5
    if (!readonly) {
      if (innerValue !== index + step) {
        setInnerValue(index + step)
      } else {
        allowClear && setInnerValue(0)
      }
    }
  }

  const getStar = (index: number, type: 'name' | 'color') => {
    let result = (type === 'name') ? voidIcon : (voidColor ?? 'void')
    const compareValue = ifHover ? hoverValue : innerValue
    if (compareValue > index) {
      result = (type === 'name') ? activeIcon : (activeColor ?? 'active')
    }
    return result
  }

  return (
    <div
      className={classNames(
        'i-rate',
        readonly && 'i-rate__readonly',
        className
      )}
      style={{ ...style }}
      onMouseEnter={handleEnterRate}
      onMouseLeave={handleLeaveRate}
      {...restProps}
    >
      {Array(count).fill('star').map((item, index) =>
        !allowHalf ? (
          <div
            className={classNames(
              'i-rate-star',
              getStar(index, 'color') === 'active' && 'i-rate-star__active',
            )}
            onMouseEnter={() => handleEnterRateItem(index)}
            onMouseDown={() => handleDownRateItem(index)}
            key={`${item}${index}`}
          >
            <Icon
              name={getStar(index, 'name')}
              color={tinycolor(getStar(index, 'color')).isValid() ? getStar(index, 'color') : undefined}
              size={20}
            />
          </div>) : (
          <div
            className='i-rate-star'
            key={`${item}${index}`}
          >
            <div
              className={classNames(
                'i-rate-star',
                'i-rate-star__first',
                getStar(index, 'color') === 'active' && 'i-rate-star__active',
              )}
              onMouseEnter={() => handleEnterRateItem(index)}
              onMouseDown={() => handleDownRateItem(index)}
            >
              <Icon
                name={getStar(index, 'name')}
                color={tinycolor(getStar(index, 'color')).isValid() ? getStar(index, 'color') : undefined}
                size={20}
              />
            </div>
            <div
              className={classNames(
                'i-rate-star',
                'i-rate-star__second',
                getStar(index, 'color') === 'active' && 'i-rate-star__active',
              )}
              onMouseEnter={() => handleEnterRateItem(index + 0.5)}
              onMouseDown={() => handleDownRateItem(index + 0.5)}
            >
              <Icon
                name={getStar(index + 0.5, 'name')}
                color={tinycolor(getStar(index + 0.5, 'color')).isValid() ? getStar(index, 'color') : undefined}
                size={20}
              />
            </div>
          </div>
        )
      )}
    </div>
  );
};

Rate.displayName = 'Rate';

export default Rate;
