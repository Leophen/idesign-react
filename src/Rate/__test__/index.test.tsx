import { render, fireEvent } from '@testing-library/react';
import Rate from '../index';

describe('Rate 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Rate />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    const { container } = render(<Rate value={3} />);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('color: rgb(245, 219, 77)');
    expect(container.firstChild.lastChild.firstChild).toHaveStyle('color: var(--i-bg-back)');
  });

  it('defaultValue', () => {
    const { container } = render(<Rate defaultValue={3} />);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('color: rgb(245, 219, 77)');
    expect(container.firstChild.lastChild.firstChild).toHaveStyle('color: var(--i-bg-back)');
  });

  it('onChange', () => {
    const changeFn = jest.fn();
    const { container } = render(<Rate onChange={changeFn} />);
    fireEvent.mouseDown(container.firstChild.firstChild);
    expect(changeFn).toHaveBeenCalled();
  });

  it('readonly', () => {
    const changeFn = jest.fn();
    const { container } = render(<Rate readonly onChange={changeFn} />);
    expect(container.firstChild).toHaveClass('i-rate__readonly');
    fireEvent.mouseDown(container.firstChild.firstChild);
    expect(changeFn).toBeCalledTimes(0);
  });

  it('allowHalf', () => {
    const { container } = render(<Rate allowHalf />);
    expect(container.querySelectorAll('.i-rate-star__first')[0]).toBeInTheDocument();
  });

  it('count', () => {
    const { container } = render(<Rate count={10} />);
    expect(container.querySelectorAll('.i-rate-star').length).toBe(10);
  });

  it('activeColor voidColor', () => {
    const { container } = render(<Rate defaultValue={3} activeColor="blue" voidColor="red" />);
    expect(container.firstChild.firstChild.firstChild).toHaveStyle('color: blue');
    expect(container.firstChild.lastChild.firstChild).toHaveStyle('color: red');
  });

  it('activeIcon voidIcon', () => {
    const { container } = render(<Rate defaultValue={3} activeIcon="BellFill" voidIcon="Bell" />);
    expect(container.firstChild.firstChild.firstChild).toHaveClass('icon-BellFill');
  });
});

