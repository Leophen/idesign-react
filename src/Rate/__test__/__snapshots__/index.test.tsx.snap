// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-rate"
  >
    <div
      class="i-rate-star"
    >
      <i
        class="i-icon icon-StarFill"
        style="font-size: 20px;"
      />
    </div>
    <div
      class="i-rate-star"
    >
      <i
        class="i-icon icon-StarFill"
        style="font-size: 20px;"
      />
    </div>
    <div
      class="i-rate-star"
    >
      <i
        class="i-icon icon-StarFill"
        style="font-size: 20px;"
      />
    </div>
    <div
      class="i-rate-star"
    >
      <i
        class="i-icon icon-StarFill"
        style="font-size: 20px;"
      />
    </div>
    <div
      class="i-rate-star"
    >
      <i
        class="i-icon icon-StarFill"
        style="font-size: 20px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
